File: web/src/strategy.ts
// EMA momentum crossover. The classic golden-cross / death-cross: a fast EMA
// crossing a slow EMA. Written ONCE against the canonical Candle shape, so it
// runs on any broker that serves candles. The math is Michael's mur EMA + Pine
// crossover semantics.
//
// No em dashes anywhere in this file.

import { Candle } from "./contract";
import { crossover, crossunder, ema } from "./indicators";

export type Stance = "BULLISH" | "BEARISH" | "FLAT";
export type Signal = "BUY" | "SELL" | "HOLD";

export interface StrategyResult {
  fast: number;
  slow: number;
  closes: number[];
  dates: string[];
  emaFast: number[];
  emaSlow: number[];
  signal: Signal; // what to do on the latest bar
  stance: Stance; // current fast-vs-slow regime
  crossIndex: number; // index of the most recent cross, or -1
  crossType: "golden" | "death" | null;
  barsSinceCross: number | null;
}

export function emaCrossover(candles: Candle[], fast = 8, slow = 21): StrategyResult {
  const closes = candles.map((c) => c.close);
  const dates = candles.map((c) => c.date);
  const emaFast = ema(closes, fast);
  const emaSlow = ema(closes, slow);

  let crossIndex = -1;
  let crossType: "golden" | "death" | null = null;
  for (let i = 1; i < closes.length; i++) {
    if (crossover(emaFast, emaSlow, i)) {
      crossIndex = i;
      crossType = "golden";
    } else if (crossunder(emaFast, emaSlow, i)) {
      crossIndex = i;
      crossType = "death";
    }
  }

  const last = closes.length - 1;
  let stance: Stance = "FLAT";
  if (!Number.isNaN(emaFast[last]) && !Number.isNaN(emaSlow[last])) {
    stance = emaFast[last] >= emaSlow[last] ? "BULLISH" : "BEARISH";
  }

  // A signal fires only on the bar the cross happens. Otherwise hold the stance.
  let signal: Signal = "HOLD";
  if (crossIndex === last) signal = crossType === "golden" ? "BUY" : "SELL";

  const barsSinceCross = crossIndex >= 0 ? last - crossIndex : null;
  return { fast, slow, closes, dates, emaFast, emaSlow, signal, stance, crossIndex, crossType, barsSinceCross };
}

File: engine/strategy.py
"""EMA momentum crossover. Fast EMA over slow EMA (golden / death cross).

Written once against the canonical Candle shape, so it runs on any broker that
serves candles. Math is Michael's mur EMA + Pine crossover semantics.

Run:  python -m engine.strategy

No em dashes anywhere in this file.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import List, Optional, Sequence

from .contract import Candle
from .indicators import crossover, crossunder, ema


@dataclass(frozen=True)
class StrategyResult:
    fast: int
    slow: int
    closes: List[float]
    ema_fast: List[float]
    ema_slow: List[float]
    signal: str  # BUY | SELL | HOLD on the latest bar
    stance: str  # BULLISH | BEARISH | FLAT
    cross_index: int  # index of the most recent cross, or -1
    cross_type: Optional[str]  # "golden" | "death" | None
    bars_since_cross: Optional[int]


def ema_crossover(candles: Sequence[Candle], fast: int = 8, slow: int = 21) -> StrategyResult:
    closes = [c.close for c in candles]
    ef = ema(closes, fast)
    es = ema(closes, slow)

    cross_index = -1
    cross_type: Optional[str] = None
    for i in range(1, len(closes)):
        if crossover(ef, es, i):
            cross_index, cross_type = i, "golden"
        elif crossunder(ef, es, i):
            cross_index, cross_type = i, "death"

    last = len(closes) - 1
    stance = "FLAT"
    if last >= 0 and not math.isnan(ef[last]) and not math.isnan(es[last]):
        stance = "BULLISH" if ef[last] >= es[last] else "BEARISH"

    signal = "HOLD"
    if cross_index == last:
        signal = "BUY" if cross_type == "golden" else "SELL"

    bars_since = (last - cross_index) if cross_index >= 0 else None
    return StrategyResult(fast, slow, closes, ef, es, signal, stance, cross_index, cross_type, bars_since)


def main():
    from . import fixtures
    from .adapters import get_adapter, supported_brokers

    raws = {
        "tradier": fixtures.RAW_TRADIER,
        "tastytrade": fixtures.RAW_TASTYTRADE,
        "snaptrade": fixtures.RAW_SNAPTRADE,
        "alpaca": fixtures.RAW_ALPACA,
        "schwab": fixtures.RAW_SCHWAB,
        "ibkr": fixtures.RAW_IBKR,
    }
    print("EMA 8/21 crossover, one strategy, every broker that serves candles:\n")
    for name in supported_brokers():
        candles = get_adapter(name, raws[name]).get_candles("AAPL")
        if not candles:
            print(f"  {name:11s} no history endpoint (aggregator), candles unsupported")
            continue
        r = ema_crossover(candles)
        when = "latest bar" if r.bars_since_cross == 0 else f"{r.bars_since_cross} bars ago"
        print(f"  {name:11s} {r.stance:8s} {r.cross_type} cross {when}  ({len(candles)} bars)")
    print("\nIdentical candles in, identical signal out. The strategy never moved.")


if __name__ == "__main__":
    main()
